ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/MergeCompactor.java

```diff
@@ -0,0 +1,224 @@
+              .matcher(fPath.getName()) : AcidUtils.BUCKET_PATTERN.matcher(fPath.getName());
+      if (!matcher.find()) {
+        String errorMessage = String
+                .format("Found a bucket file which did not match the bucket pattern! %s Matcher=%s", fPath,
+                        matcher);
+        LOG.error(errorMessage);
+        throw new IllegalArgumentException(errorMessage);
+      }
+      int bucketNum = matcher.groupCount() > 0 ? Integer.parseInt(matcher.group(1)) : Integer.parseInt(matcher.group());
+      bucketIdToBucketFilePath.computeIfAbsent(bucketNum, ArrayList::new);
+      Reader reader = OrcFile.createReader(fs, fPath);
+      bucketIdToBucketFilePath.computeIfPresent(bucketNum, (k, v) -> v).add(reader);
+    }
+    return bucketIdToBucketFilePath;
+  }
+
+  /**
+   * Generate output path for compaction. This can be used to generate delta or base directories.
+   * @param conf hive configuration, must be non-null
+   * @param writeIds list of valid write IDs
+   * @param isBaseDir if base directory path should be generated
+   * @param sd the resolved storadge descriptor
+   * @return output path, always non-null
+   */
+  private Path getOutputDirPath(HiveConf conf, ValidWriteIdList writeIds, boolean isBaseDir,
+                                         StorageDescriptor sd) {
+    long minOpenWriteId = writeIds.getMinOpenWriteId() == null ? 1 : writeIds.getMinOpenWriteId();
+    long highWatermark = writeIds.getHighWatermark();
+    long compactorTxnId = Compactor.getCompactorTxnId(conf);
+    AcidOutputFormat.Options options = new AcidOutputFormat.Options(conf).writingBase(isBaseDir)
+            .writingDeleteDelta(false).isCompressed(false).minimumWriteId(minOpenWriteId)
+            .maximumWriteId(highWatermark).statementId(-1).visibilityTxnId(compactorTxnId);
+    return AcidUtils.baseOrDeltaSubdirPath(new Path(sd.getLocation()), options);
+  }
+
+  /**
+   * Merge ORC files from base/delta directories. If the directories contains multiple buckets, the result will also
+   * contain the same amount.
+   * @param conf hive configuration
+   * @param includeBaseDir if base directory should be scanned for orc files
+   * @param dir the root directory of the table/partition
+   * @param outputDirPath the result directory path
+   * @param isMm merge orc files from insert only tables
+   * @throws IOException error occurred during file operation
+   */
+  private boolean mergeFiles(HiveConf conf, boolean includeBaseDir, AcidDirectory dir,
+                               Path outputDirPath, boolean isMm) throws IOException {
+    Map<Integer, List<Reader>> bucketIdToBucketFiles = matchBucketIdToBucketFiles(conf, dir, includeBaseDir, isMm);
+    OrcFileMerger fileMerger = new OrcFileMerger(conf);
+    boolean isCompatible = true;
+    for (Map.Entry<Integer, List<Reader>> e : bucketIdToBucketFiles.entrySet()) {
+      isCompatible &= fileMerger.checkCompatibility(e.getValue());
```

Comment:
should we refactor this to find the first incompatible and return?
````
bucketIdToBucketFiles.values.stream().allMatch(v->fileMerger.checkCompatibility(v))
````

Reply:
Nice way to refactor @deniskuzZ . Done.